### PR TITLE
124 3.21 age when the event starts

### PR DESCRIPTION
**Pull request by frederike-ramin**
Show the applicant's age when the event starts, not the current age. (On events details, application overview page and on generated pdf)

Take your time for reviewing and merry christmas^^

### CODE REVIEW

File: config/locales/de.events.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ de:
       filter: "Filtern"
       status: "Status"
       details: "Details"
+      age_when_event_starts: "Alter <br> bei Beginn"
```
Comment (bjrne) [inline config/locales/de.events.yml]: I doubt this is the right time for formatting

File: app/models/application_letter.rb [hunk 2]
```diff
@@ -83,4 +83,15 @@ def status_cannot_be_changed
...
+
+  # Returns the age of the user based on the date the event starts
+  #
+  # @param none
+  # @return [Int] for age as number of years
+  def applicant_age_when_event_starts
+    start_date = event.start_date
+    birth_date = user.profile.birth_date
+    start_date_is_before_birthday = start_date.month > birth_date.month || (start_date.month == birth_date.month && start_date.day >= birth_date.day)
+    return start_date.year - birth_date.year - (start_date_is_before_birthday ? 0 : 1)
```
Comment (bjrne) [inline app/models/application_letter.rb]: I don't have a better idea at the moment, but maybe someone knows how to improve the readability of these two lines

**Review by bjrne — COMMENTED** [review]
Rest seems good 👍

File: app/models/application_letter.rb (hunk 2 above)
Comment (querenker) [inline app/models/application_letter.rb]: This method is very similar to [User#older_than_required_age_at_start_date_of_event?](https://github.com/hpi-swt2/workshop-portal/blob/dev/app/models/user.rb#L80) and [Profile#age](https://github.com/hpi-swt2/workshop-portal/blob/dev/app/models/profile.rb#L33). I think it should be a good idea to provide this functionality in a module or something like this?

File: app/views/events/_applicants_overview.html.erb [hunk 3]
```diff
@@ -34,13 +34,14 @@
     </thead>
     <tbody>
     <%
-      @application_letters.sort_by! {|l| l.user.profile.send(params[:sort]) } if params[:sort]
+      @application_letters.sort_by! {|l| l.user.profile.send(params[:sort]) } if params[:sort] && params[:sort] != 'applicant_age_when_event_starts'
+      @application_letters.sort_by! {|l| l.send(params[:sort]) } if params[:sort] && params[:sort] == 'applicant_age_when_event_starts'
       @application_letters.reverse! if params[:order] == 'descending'
       @application_letters.select {|l| l.user.profile }.each do | application_letter | %>
```
Comment (querenker) [inline app/views/events/_applicants_overview.html.erb]: I know this was already there and is probably not part of this user story. But I think this code should not be in the view?

File: spec/models/application_letter_spec.rb [hunk 4]
```diff
@@ -67,4 +67,12 @@
      application.status = :rejected
      expect(application).to be_valid
   end
+
+  it "calculates the correct age of applicant when event starts" do
```
Comment (querenker) [inline spec/models/application_letter_spec.rb]: There are some edge cases which are not covered in this test. One more reason to factor this functionality out :)

**Review by querenker — CHANGES_REQUESTED** [review]
Otherwise it looks good to me :)

File: app/views/events/_applicants_overview.html.erb (hunk 3 above)
Comment (frederike-ramin) [inline app/views/events/_applicants_overview.html.erb]: Though, the sorting needs to be done on client-side. So the only option is view, right?

File: config/locales/de.events.yml (hunk 1 above)
Comment (frederike-ramin) [inline config/locales/de.events.yml]: Yeah..I wasn't sure either^^
But I'm also not very sure if the fix is that much better..

File: app/views/events/_applicants_overview.html.erb (hunk 3 above)
Comment (MasterCarl) [inline app/views/events/_applicants_overview.html.erb]: I wrote this line, and I do think that it should probably go in the controller (although it's more of a visual aspect than business logic). Since view rendering is server-side, this could be in either place (controller or view).

File: app/views/events/_applicants_overview.html.erb [hunk 5]
```diff
@@ -13,7 +13,7 @@
         <%= sort_caret(Profile.human_attribute_name(:gender), :gender) %>
       </th>
       <th>
-        <%= sort_caret(Profile.human_attribute_name(:age), :age) %>
+        <%= sort_caret(( t('.age') + " <br> " + t('.when_event_starts')), :applicant_age_when_event_starts ) %>
```
Comment (querenker) [inline app/views/events/_applicants_overview.html.erb]: I think the problem is the <br> tag in general. The HTML should describe the structure of the document independent of the visual representation. There are some cases where a line break has semantic meaning like in adresses or poetries. But I think this isn't such a case and therefore the positioning should be handled with CSS files?
I don't know how to archive this with CSS myself, but probably there exists some CSS magic ;) For now I think it isn't that bad that it is on one line if the room is available. Otherwise it should break the line either way?

Comment (frederike-ramin) [inline app/views/events/_applicants_overview.html.erb]: It doesn't break the line if you don't do it maually. Thing is, that it fits into the line, but that makes the columns besides that one smaller which leads to a linebreak in the status-update-toggle-fields. You know what I'm talking about?

File: app/models/profile.rb [hunk 6]
```diff
@@ -26,14 +26,13 @@ def adult?()
     self.birth_date >= 18.years.ago
   end
 
-  # Returns the age of the user based on the current date
+  # Returns the age of the user based on the given date
   #
   # @param none
   # @return [Int] for age as number of years
-  def age
```
Comment (cmfcmf) [inline app/models/profile.rb]: I would've preferred to keep the `age` method:
```ruby
def age
   return age_at_time(Time.now)
end
```

File: app/views/events/_applicants_overview.html.erb (hunk 5 above)
Comment (cmfcmf) [inline app/views/events/_applicants_overview.html.erb]: A simple `style="max-width: 150px"` on the enclosing `<th>` can fix it :)

Comment (frederike-ramin) [inline app/views/events/_applicants_overview.html.erb]: @cmfcmf You. Are. An. Angel. Thanks!

File: config/locales/de.events.yml [hunk 7]
```diff
@@ -10,6 +10,7 @@ de:
       filter: "Filtern"
       status: "Status"
       details: "Details"
+      age_when_event_starts: "Alter bei Beginn"
```
Comment (querenker) [inline config/locales/de.events.yml]: There are two identity versions of this local. Maybe one can be replaced?

**Review by querenker — CHANGES_REQUESTED** [review]

File: config/locales/de.events.yml (hunk 7 above)
Comment (frederike-ramin) [inline config/locales/de.events.yml]: They have different contexts. It wouldn't make sense to join them together.

**Review by querenker — APPROVED** [review]

**Review by aloeser — APPROVED** [review]
works for me locally

**Review by ekrebs5 — APPROVED** [review]
